src/Uno.UI/UI/Xaml/Controls/SwipeControl/SwipeControl.cs

```diff
@@ -0,0 +1,1893 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+// Imported in uno on 2021/03/21 from commit 307bd99682cccaa128483036b764c0b7c862d666
+// https://github.com/microsoft/microsoft-ui-xaml/blob/307bd99682cccaa128483036b764c0b7c862d666/dev/SwipeControl/SwipeControl.cpp
+
+using System;
+using System.Numerics;
+using Windows.Foundation;
+using Windows.Foundation.Collections;
+using Windows.UI.Composition;
+using Windows.UI.Core;
+using Windows.UI.Xaml.Hosting;
+using Windows.UI.Xaml.Input;
+using Windows.UI.Xaml.Media;
+using Uno.Disposables;
+using Uno.UI.Helpers.WinUI;
+using Uno.UI.Extensions;
+using System.Threading.Tasks;
+
+namespace Windows.UI.Xaml.Controls
+{
+	public partial class SwipeControl : ContentControl
+	{
+
+		// Change to 'true' to turn on debugging outputs in Output window
+		//bool SwipeControlTrace.s_IsDebugOutputEnabled = false;
+		//bool SwipeControlTrace.s_IsVerboseDebugOutputEnabled = false;
+
+		private const double c_epsilon = 0.0001;
+		private const float c_ThresholdValue = 100.0f;
+		private const float c_MinimumCloseVelocity = 31.0f;
+
+		[ThreadStatic]
+		static WeakReference<SwipeControl> s_lastInteractedWithSwipeControl = null;
+
+		public SwipeControl()
+		{
+			s_lastInteractedWithSwipeControl ??= new WeakReference<SwipeControl>(null);
+
+			//__RP_Marker_ClassById(RuntimeProfiler.ProfId_SwipeControl);
+			SetDefaultStyleKey(this);
+
+			Loaded += UnfinalizeOnLoad;
+			Unloaded += FinalizeOnUnload;
+		}
+
+		// [BEGIN] Uno workaround:
+		//  + we make sure to un-subscribe from events on unload to avoid leaks
+		//  + we must not use finalizer on uno (invoked from a bg thread)
+
+		private static void UnfinalizeOnLoad(object sender, RoutedEventArgs routedEventArgs)
+			=> ((SwipeControl)sender).SwipeControl_Unfinalizer();
+
+		private static void FinalizeOnUnload(object sender, RoutedEventArgs routedEventArgs)
+			=> ((SwipeControl)sender).SwipeControl_Finalizer();
+
+		private void SwipeControl_Unfinalizer()
+		{
+			DetachEventHandlers(); // Do not double subscribe
+			AttachEventHandlers(isUnoUnfinalizer: true);
+		}
+		
+		private void SwipeControl_Finalizer()
+		//~SwipeControl()
+		// [END] Uno workaround
+		{
+			DetachEventHandlers();
+
+			if (s_lastInteractedWithSwipeControl.TryGetTarget(out var lastInteractedWithSwipeControl))
+			{
+				if (lastInteractedWithSwipeControl == this)
+				{
+					s_lastInteractedWithSwipeControl.SetTarget(null);
+					var globalTestHooks = SwipeTestHooks.GetGlobalTestHooks();
+					if (globalTestHooks is {})
+					{
+						globalTestHooks.NotifyLastInteractedWithSwipeControlChanged();
+					}
+				}
+			}
+		}
+
+		#region ISwipeControl
+		public async void Close()
```

Comment:
Async void is likely to cause race conditions when calling it multiple times. We should wrap it and keep track so we at least queue operations properly.